internal/alb/sg/association.go

```diff
@@ -0,0 +1,259 @@
+	return controller.deletedManagedSGs(association)
+}
+
+func (controller *associationController) reconcileWithExternalSGs(association *Association) error {
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: association.ExternalSGIDs,
+		LbArn:    association.LbArn,
+	}
+	err := controller.lbAttachmentController.Reconcile(lbSGAttachment)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile external LoadBalancer securityGroup due to %s", err.Error())
+	}
+
+	err = controller.deletedManagedSGs(association)
+	if err != nil {
+		return fmt.Errorf("failed to delete managed securityGroups due to %s", err.Error())
+	}
+	return nil
+}
+
+func (controller *associationController) reconcileWithManagedSGs(association *Association) error {
+	lbSG, err := controller.reconcileManagedLbSG(association)
+	if err != nil {
+		return err
+	}
+	err = controller.reconcileManagedInstanceSG(association, lbSG)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (controller *associationController) reconcileManagedLbSG(association *Association) (*SecurityGroup, error) {
+	lbSGName := controller.namer.NameLbSG(association.LbID)
+	lbSG := &SecurityGroup{
+		GroupName: &lbSGName,
+	}
+	for _, port := range association.LbPorts {
+		ipRanges := []*ec2.IpRange{}
+		for _, cidr := range association.LbInboundCIDRs {
+			ipRanges = append(ipRanges, &ec2.IpRange{
+				CidrIp:      cidr,
+				Description: aws.String(fmt.Sprintf("Allow ingress on port %v from %v.", port, aws.StringValue(cidr))),
+			})
+		}
+		permission := &ec2.IpPermission{
+			IpProtocol: aws.String("tcp"),
+			FromPort:   aws.Int64(port),
+			ToPort:     aws.Int64(port),
+			IpRanges:   ipRanges,
+		}
+		lbSG.InboundPermissions = append(lbSG.InboundPermissions, permission)
+	}
+
+	err := controller.sgController.Reconcile(lbSG)
+	if err != nil {
+		return lbSG, fmt.Errorf("failed to reconcile managed LoadBalancer securityGroup due to %s", err.Error())
+	}
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: []string{*lbSG.GroupID},
+		LbArn:    association.LbArn,
+	}
+	err = controller.lbAttachmentController.Reconcile(lbSGAttachment)
+	if err != nil {
+		return lbSG, fmt.Errorf("failed to reconcile managed LoadBalancer securityGroup attachment due to %s", err.Error())
+	}
+	return lbSG, nil
+}
+
+func (controller *associationController) reconcileManagedInstanceSG(association *Association, lbSG *SecurityGroup) error {
+	instanceSGName := controller.namer.NameInstanceSG(association.LbID)
+	instanceSG := &SecurityGroup{
+		GroupName: &instanceSGName,
+		InboundPermissions: []*ec2.IpPermission{
+			{
+				IpProtocol: aws.String("tcp"),
+				FromPort:   aws.Int64(0),
+				ToPort:     aws.Int64(65535),
+				UserIdGroupPairs: []*ec2.UserIdGroupPair{
+					{
+						GroupId: lbSG.GroupID,
+					},
+				},
+			},
+		},
+	}
+	err := controller.sgController.Reconcile(instanceSG)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile managed Instance securityGroup due to %s", err.Error())
```

Comment:
NIT. The `.Error()` at the end is redundant. It should still print the error message without calling the `Error()` method explicitly.

Reply:
i'll change the errors to be "XXX due to %v", err